### PR TITLE
feat(imgproc): Lanczos-3 GPU kernels for resize and warp-affine

### PR DESCRIPTION
**Pull request by Incharajayaram**
## 📝 Description

**Fixes/Relates to:** GSoC 2026

Adds Lanczos-3 interpolation GPU kernels for both resize and warp-affine operations.

**Changes:**
- `launch_resize_lanczos_cuda`: separable 2-pass Lanczos-3 (horizontal 6-tap → vertical 6-tap through a `dst_w × src_h` intermediate buffer), reducing source reads from 36 to 12 vs a naïve 2D kernel
- `launch_warp_affine_lanczos_cuda`: full 6×6 2D Lanczos-3 for warp-affine (arbitrary affine maps cannot be decomposed into separable passes)
- Both kernels use `__sinf`/`__fdividef` fast intrinsics, normalized weights (boundary-safe), `#pragma unroll`, `fmaf`, and `__ldg` read-only cache loads
- Extended `bench_gpu_resize` and `bench_gpu_warp_affine` Rust examples with Lanczos sections
- Extended `bench_opencv_resize.py` and `bench_opencv_warp_affine.py` with Lanczos comparison sections

---

## 🛠️ Changes Made
- [x] `crates/kornia-imgproc/src/gpu/resize_cuda.rs` — separable 2-pass Lanczos-3 resize kernel + launcher
- [x] `crates/kornia-imgproc/src/gpu/warp_affine_cuda.rs` — 2D 6×6 Lanczos-3 warp-affine kernel + launcher
- [x] `crates/kornia-imgproc/examples/bench_gpu_resize.rs` — Lanczos benchmark section
- [x] `crates/kornia-imgproc/examples/bench_gpu_warp_affine.rs` — Lanczos benchmark section
- [x] `crates/kornia-imgproc/examples/bench_opencv_resize.py` — OpenCV CPU/CUDA Lanczos comparison
- [x] `crates/kornia-imgproc/examples/bench_opencv_warp_affine.py` — OpenCV CPU/CUDA Lanczos comparison

---

## 🧪 How Was This Tested?

- [x] **Manual Verification:** Run on GTX 1650 (sm_75), 50 warmup + 200 timed iterations, 3-channel f32 images

### Resize — separable 2-pass (GTX 1650)

| Case | kornia-rs Lanczos-3 | cv2 CPU Lanczos-4 | Speedup |
|------|---------------------|-------------------|---------|
| 1080p → 540p | 0.402 ms | 3.884 ms | **9.7×** |
| 1080p → 4K   | 1.995 ms | ~18 ms   | **9×**  |
| 4K → 1080p   | 1.592 ms | ~12 ms   | **7.5×** |

### Warp-affine — 2D kernel (45° rotation, GTX 1650)

| Case | kornia-rs Lanczos-3 | cv2 CPU Lanczos-4 |
|------|---------------------|-------------------|
| 1920×1080 | 4.262 ms | ~43–51 ms (**10–12× faster**) |

### Ecosystem comparison

`cv2.cuda.resize` and `cv2.cuda.warpAffine` error on `INTER_LANCZOS4` (only NEAREST/LINEAR/CUBIC/AREA supported). `torch.nn.functional.interpolate` has no Lanczos mode. **kornia-rs is the only library providing GPU Lanczos interpolation in this comparison.**

- [x] **Performance/Edge Cases:** Normalized weights prevent brightness drift at OOB-clamped boundaries

---

## 🕵️ AI Usage Disclosure
- [x] 🟡 **AI-assisted:** I used AI for boilerplate/refactoring but have manually reviewed and tested every line.

---

## 🚦 Checklist
- [x] My code follows the existing style guidelines of this project.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have performed a **self-review** of my code (no "ghost" variables or hallucinations).

---

## 💭 Additional Context

The separable 2-pass design for resize is the key optimization: instead of 36 reads per output pixel (6×6 2D), pass 1 writes a `dst_w × src_h` intermediate buffer (6 reads/pixel horizontal), and pass 2 reads from that intermediate (6 reads/pixel vertical). For 2× upscale this saves 4× reads; for 2× downscale it saves 2× reads — matching the measured speedup.

Warp-affine stays 2D because arbitrary affine transforms (rotation, shear) scatter source coordinates and cannot be decomposed.

### CODE REVIEW

**Review by edgarriba — COMMENTED** [review]
## Review — Lanczos-3 GPU kernels

Checked the kernel math carefully (Lanczos sinc formula `3·sin(πx)·sin(πx/3)/(π²x²)`, tap↔weight pairing `x0-2..x0+3` vs `L(frac+2)..L(frac-3)`, weight normalization, boundary clamping, intermediate buffer sizing/strides, and pass-1/pass-2 dimension wiring) — all correct. The fixed 6-tap window aliases on strong downscale, but that intentionally matches OpenCV `INTER_LANCZOS4` (the CPU path's `antialias=false` mode), so it's consistent, not a bug. No correctness defect found.

### Finding — `crates/kornia-imgproc/src/cuda/resize.rs` (`launch_resize_lanczos_cuda`)

The intermediate is allocated **and zeroed** on every call:

```rust
let mut intermediate = stream.alloc_zeros::<f32>(inter_len)?;
```

Pass 1 (`resize_lanczos_h_3c`) writes every element of the `dst_w × src_h × 3` buffer (≈50 MB for the 1080p↔4K cases) before pass 2 reads it, so the zero-fill is pure waste — roughly ~0.15–0.3 ms against a ~2 ms kernel (~10–15% overhead). The per-call `cudaMalloc`/free also churns the allocator on tight resize loops, whereas every other resize launcher in this module is zero-alloc.

Suggest an uninitialized allocation and, ideally, a caller-provided/cached scratch buffer. (The doc comment already acknowledges the per-call alloc.)

### Minor / non-blocking
- `lanczos3_h` / `lanczos3_v` / `lanczos3_wa` are three byte-identical device functions — unavoidable across separate NVRTC modules, but the `_h`/`_v` split within `resize.rs` is a duplication smell.
- `launch_resize_lanczos_cuda` omits the `block_dim: Option<(u32,u32)>` parameter that every other resize launcher exposes. Harmless (auto-clamp still works) but an API inconsistency.

_Reviewed with `/review` (medium effort)._
